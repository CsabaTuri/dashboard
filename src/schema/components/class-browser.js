import React, { Component } from "react"
import { connect } from "react-redux"
import { query as q } from "faunadb"
import { Link } from "react-router"
import { Button, ButtonType } from "office-ui-fabric-react"

import { selectedClass, selectedDatabase, createIndex } from "../"
import { faunaClient } from "../../authentication"
import { isBusy, monitorActivity } from "../../activity-monitor"
import { watchForError } from "../../notifications"
import { linkForRef } from "../../router"
import { InstancesList } from "../../dataset"
import { KeyType } from "../../persistence/faunadb-wrapper"

export class ClassBrowser extends Component {

  constructor(props) {
    super(props)
    this.state = this.initialState(props)
    this.createClassIndex = this.createClassIndex.bind(this)
    this.onSelectRef = this.onSelectRef.bind(this)
  }

  initialState(props) {
    const index = this.findClassCoveringRef(props.clazz)

    return  {
      index,
      queryFn: this.buildQueryFn(props, index)
    }
  }

  findClassCoveringRef(clazz) {
    const res = clazz.get("indexes")
      .filter(idx =>
        idx.get("terms").isEmpty() && (
          idx.get("values").isEmpty() ||
          idx.get("values").some(v => v.getIn(["field", 0]) === "ref")
        )
      )
      .sort((a, b) => {
        if (a.get("values").isEmpty() && b.get("values").isEmpty())
          return a.get("name") < b.get("name") ? -1 : 1

        if (a.get("values").isEmpty()) return -1
        if (b.get("values").isEmpty()) return 1
        return 0
      })

    return res.isEmpty() ? null : res.get(0)
  }

  componentDidMount() {
    this.setState(this.initialState(this.props))
  }

  componentWillReceiveProps(next) {
    if (this.props.clazz !== next.clazz) {
      this.setState(this.initialState(next))
    }
  }

  buildQueryFn({ client, path }, index) {
    return opts =>
      client.query(path, KeyType.SERVER,
        q.Map(
          q.Paginate(q.Match(index.get("ref")), opts),
          res => this.getRef(res, index)
        )
      ).then(res => ({
        ...res,
        data: res.data.map(inst => ({
          ref: inst.ref,
          ...inst.data
        }))
      }))
  }

  getRef(res, index) {
    const values = index.get("values")
    if (values.isEmpty()) return q.Get(res)
    if (values.size === 1) return q.Get(res)

    return q.Get(q.Select(
      values
        .zip(values.keySeq())
        .find(v => v[0].getIn(["field", 0]) === "ref")[1],
      res
    ))
  }

  createClassIndex(e) {
    e.preventDefault()
    const { client, path, clazz } = this.props

    return this.props.dispatch(
      monitorActivity(
        watchForError(
          "Unexpected error while creating class index",
          createIndex(client, path, {
            name: `all_${clazz.get("name")}`,
            source: clazz.get("ref")
          })
        )
      )
    )
  }

  onSelectRef(ref) {
    const { client, path } = this.props

    return this.props.dispatch(
      monitorActivity(
        watchForError(
          "Unexpected error while fetching instance information", () =>
          client.query(path, KeyType.SERVER, q.Get(ref))
        )
      )
    )
  }

  render() {
    if (!this.state.index) {
      return this.renderNoClassIndexPage()
    }

    const { databaseUrl } = this.props
    const { index } = this.state
    const indexLink = linkForRef(databaseUrl, index.get("ref"))

    return <div>
      <InstancesList query={this.state.queryFn} onSelectRef={this.onSelectRef} />
      <p className="ms-TextField-description">
        This view was derived from <Link to={indexLink.get("url")}>{indexLink.get("name")}</Link>.
      </p>
    </div>
  }

  renderNoClassIndexPage() {
    const { isBusy } = this.props

    return <div>
      <p>No eligible class index found.</p>
      <p className="ms-TextField-description">
        Without a class index instances can only be loaded by
        Ref. A class index indexes all members of the class under a single key. For large
        datasets this can increase storage and processing overhead, so use class
        indexes sparingly in production.
      </p>
      <Button
        buttonType={ButtonType.primary}
        disabled={isBusy}
        onClick={this.createClassIndex}>
          Create a class index now
      </Button>
    </div>
  }

}

export default connect(
  state => ({
    path: selectedDatabase(state).get("path"),
    databaseUrl: selectedDatabase(state).get("url"),
    clazz: selectedClass(state),
    client: faunaClient(state),
    isBusy: isBusy(state)
  })
)(ClassBrowser)
