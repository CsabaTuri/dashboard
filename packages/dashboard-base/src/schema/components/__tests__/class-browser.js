import React from "react"
import { shallow } from "enzyme"
import { shallowToJson } from "enzyme-to-json"
import { query as q } from "faunadb"
import { values as v } from "faunadb"
import { Map, List } from "immutable"

import { ClassBrowser } from "../class-browser"

jest.mock("../../../notifications", () => ({ watchForError: (msg, fn) => fn() }))
jest.mock("../../../activity-monitor", () => ({ monitorActivity: jest.fn() }))
jest.mock("../../", () => ({ createIndex: jest.fn(() => () => null) }))

const { createIndex } = require("../../")

describe("ClassBrowser Component", () => {
  let comp,
    classWithClassIndex,
    classWithNonClassIndex,
    classWithoutIndex,
    client

  beforeEach(() => {
    client = {
      query: jest.fn(() => Promise.resolve({
        data: [{
          ref: new v.Ref("1", new v.Ref("people", v.Native.CLASSES)),
          data: { name: "Bob" }
        }]
      }))
    }

    classWithClassIndex = Map.of(
      "name", "people",
      "ref", new v.Ref("people", v.Native.CLASSES),
      "indexes", List.of(
        Map.of(
          "name", "all_people",
          "ref", new v.Ref("all_people", v.Native.INDEXES),
          "terms", List(),
          "values", List()
        )
      )
    )

    classWithNonClassIndex = Map.of(
      "name", "people",
      "ref", new v.Ref("people", v.Native.CLASSES),
      "indexes", List.of(
        Map.of(
          "name", "all_people",
          "ref", new v.Ref("all_people", v.Native.INDEXES),
          "terms", List(),
          "values", List.of(
            Map.of("field", List.of("data", "name"))
          )
        )
      )
    )

    classWithoutIndex = Map.of(
      "name", "users",
      "ref", new v.Ref("users", v.Native.CLASSES),
      "indexes", List.of()
    )

    comp = shallow(
      <ClassBrowser
        dispatch={jest.fn(x => x)}
        isBusy={false}
        client={client}
        clazz={classWithClassIndex}
        databaseUrl="/db/a-db/databases"
        path={["a-db"]} />
    )
  })

  it("displays instances", () => {
    expect(shallowToJson(comp)).toMatchSnapshot()
  })

  it("should respond to changes in pagination", () => {
    comp.find("InstancesList").props().query({ size: 1 })
    expect(client.query).toHaveBeenCalled()
  })

  it("should respond to selected ref", () => {
    comp.find("InstancesList").simulate("selectRef", "a-ref")
    expect(client.query).toHaveBeenCalled()
  })

  it("should display new index button when no class index", () => {
    comp.setProps({ clazz: classWithoutIndex })
    expect(shallowToJson(comp)).toMatchSnapshot()
  })

  it("should disable create index button when busy", () => {
    comp.setProps({ clazz: classWithoutIndex, isBusy: true })
    expect(shallowToJson(comp)).toMatchSnapshot()
  })

  it("should be able to create a new class index", () => {
    comp.setProps({ clazz: classWithoutIndex })
    comp.find("ComponentWithInjectedProps").simulate("click", { preventDefault: jest.fn() })

    expect(createIndex).toHaveBeenLastCalledWith(client, ["a-db"], {
      name: "all_users",
      source: new v.Ref("users", v.Native.CLASSES)
    })
  })

  it("should find a non conflicting name for new index", () => {
    comp.setProps({ clazz: classWithNonClassIndex })
    comp.find("ComponentWithInjectedProps").simulate("click", { preventDefault: jest.fn() })

    expect(createIndex).toHaveBeenLastCalledWith(client, ["a-db"], {
      name: "all_people_1",
      source: new v.Ref("people", v.Native.CLASSES)
    })
  })
})
