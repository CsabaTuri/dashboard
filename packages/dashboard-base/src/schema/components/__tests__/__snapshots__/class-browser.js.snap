// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassBrowser Component displays instances 1`] = `
<div>
  <InstancesList
    onSelectRef={[Function]}
    query={[Function]}
  />
  <p
    className="ms-TextField-description"
  >
    This view was derived from 
    <Link
      onlyActiveOnIndex={false}
      style={Object {}}
      to="/db/a-db/indexes/all_people"
    >
      indexes/all_people
    </Link>
    .
  </p>
</div>
`;

exports[`ClassBrowser Component should disable create index button when busy 1`] = `
<div>
  <p>
    No eligible class index found.
  </p>
  <p
    className="ms-TextField-description"
  >
    Without a class index instances can only be loaded by Ref. A class index indexes all members of the class under a single key. For large datasets this can increase storage and processing overhead, so use class indexes sparingly in production.
  </p>
  <ComponentWithInjectedProps
    disabled={true}
    onClick={[Function]}
  >
    Create a class index now
  </ComponentWithInjectedProps>
</div>
`;

exports[`ClassBrowser Component should display new index button when no class index 1`] = `
<div>
  <p>
    No eligible class index found.
  </p>
  <p
    className="ms-TextField-description"
  >
    Without a class index instances can only be loaded by Ref. A class index indexes all members of the class under a single key. For large datasets this can increase storage and processing overhead, so use class indexes sparingly in production.
  </p>
  <ComponentWithInjectedProps
    disabled={false}
    onClick={[Function]}
  >
    Create a class index now
  </ComponentWithInjectedProps>
</div>
`;
